docs_src/design/adr/0023-North-South-Messaging.md

```diff
@@ -0,0 +1,527 @@
+
+#### device service configuration
+
+Example device service configuration is provided below.
+
+``` toml
+[MessageQueue]
+## already existing message queue configuration (for sending events/readings to the message bus)
+Protocol = "redis"
+Host = "localhost"
+Port = 6379
+Type = "redis"
+AuthMode = "usernamepassword"                           # required for redis messagebus (secure or insecure).
+SecretName = "redisdb"
+PublishTopicPrefix = "edgex/events/device" # /<device-profile-name>/<device-name>/<source-name> will be added to this Publish Topic prefix
+  [MessageQueue.Optional]
+  # Default MQTT Specific options that need to be here to enable environment variable overrides of them
+  # Client Identifiers
+  ClientId = "device-rest"
+  # Connection information
+  Qos = "0" # Quality of Service values are 0 (At most once), 1 (At least once) or 2 (Exactly once)
+  KeepAlive = "10" # Seconds (must be 2 or greater)
+  Retained = "false"
+  AutoReconnect = "true"
+  ConnectTimeout = "5" # Seconds
+  SkipCertVerify = "false" # Only used if Cert/Key file or Cert/Key PEMblock are specified
+
+## new configuration to allow device services to also communicate via message bus with core command
+CommandRequestTopic = "edgex/command/request/#"         # subscribing for inbound command requests
+CommandResponseTopicPrefix = "edgex/command/response/"  # publishing outbound command responses; <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
+```
+
+!!! Note
+        Most of the device service configuration is existing based on its need to already communicate with the message bus for publishing events/readings.  The last two lines are added to allow device services to subscribe and publish command messages from/to the message bus.
+
+## EdgeX Service (Internal) Message Bus Requests
+
+Application services (or other EdgeX services in the future) may want to also use message communications to make command requests.  Application services make command requests today via REST.
+
+In order to support this, the following need to be added:
+
+- The command service will also need an internal request topic and internal response topic prefix configuration to allow internal EdgeX services to make command requests (and query requests).
+
+    ``` toml
+        [MessageQueue]
+            [InternalMessageQueue]
+            Protocol = "redis"
+            Host = "localhost"
+            Port = 6379
+            Type = "redis"
+            RequestTopicPrefix = "edgex/command/request/"   # for publishing requests to the device service; <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
+            ResponseTopic = "edgex/command/response/#”      # for subscribing to device service responses
+            InternalRequestCommandTopic = "/command/request/#"     # for subscribing to internal command requests
```

Comment:
Do the internal topics really need to be configurable?

Reply:
I guess they wouldn't have to be, but then again neither would the other topics we are using.  Will bring this up at TSC on 4/27.